File: Greedy Algorithms and Their Applications/Coin Change Problem (Greedy Algorithm)/Coin-Change-Problem.c
#include <stdio.h>

void greedy_coin_change(int denominations[], int n, int amount) {
    printf("Coins used: ");
    for (int i = 0; i < n; i++) {
        while (amount >= denominations[i]) {
            amount -= denominations[i];
            printf("%d ", denominations[i]);
        }
    }
    printf("\n");
}

// Sorting function (Descending Order)
void sort_desc(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = i + 1; j < n; j++) {
            if (arr[i] < arr[j]) {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
    }
}

int main() {
    int n, amount;
    printf("Enter the number of denominations: ");
    scanf("%d", &n);
    
    int denominations[n];
    printf("Enter the denominations: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &denominations[i]);
    }

    printf("Enter the amount: ");
    scanf("%d", &amount);

    sort_desc(denominations, n);
    greedy_coin_change(denominations, n, amount);

    return 0;
}

File: Greedy Algorithms and Their Applications/Coin Change Problem (Greedy Algorithm)/Coin-Change-Problem.java
import java.util.Arrays;
import java.util.Scanner;

public class GreedyCoinChange {
    public static void greedyCoinChange(int[] denominations, int amount) {
        Arrays.sort(denominations);
        reverseArray(denominations); // Sorting in descending order
        
        System.out.print("Coins used: ");
        for (int coin : denominations) {
            while (amount >= coin) {
                amount -= coin;
                System.out.print(coin + " ");
            }
        }
        System.out.println();
    }

    // Helper function to reverse an array (Descending Order)
    public static void reverseArray(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the number of denominations: ");
        int n = scanner.nextInt();
        
        int[] denominations = new int[n];
        System.out.print("Enter the denominations: ");
        for (int i = 0; i < n; i++) {
            denominations[i] = scanner.nextInt();
        }

        System.out.print("Enter the amount: ");
        int amount = scanner.nextInt();

        greedyCoinChange(denominations, amount);
        
        scanner.close();
    }
}
